l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,678 @@
+    modifier onlyBridgehubOrEra(uint256 _chainId) {
+        require(
+            (msg.sender == address(bridgehub)) ||
+                ((_chainId == ERA_CHAIN_ID) && (msg.sender == bridgehub.getStateTransition(_chainId))),
+            "L1SharedBridge: not bridgehub or era chain"
+        );
+        _;
+    }
+
+    /// @notice Checks that the message sender is the legacy bridge
+    modifier onlyLegacyBridge() {
+        require(msg.sender == address(legacyBridge), "ShB not legacy bridge");
+        _;
+    }
+
+    /// @dev Contract is expected to be used as proxy implementation.
+    /// @dev Initialize the implementation to prevent Parity hack.
+    constructor(
+        address payable _l1WethAddress,
+        IBridgehub _bridgehub,
+        IL1ERC20Bridge _legacyBridge
+    ) reentrancyGuardInitializer {
+        l1WethAddress = _l1WethAddress;
+        bridgehub = _bridgehub;
+        legacyBridge = _legacyBridge;
+    }
+
+    /// @dev Initializes a contract bridge for later use. Expected to be used in the proxy
+    /// @param _owner Address which can change L2 token implementation and upgrade the bridge
+    /// implementation. The owner is the Governor and separate from the ProxyAdmin from now on, so that the Governor can call the bridge
+    function initialize(
+        address _owner,
+        uint256 _eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber
+    ) external reentrancyGuardInitializer reinitializer(2) {
+        _transferOwnership(_owner);
+        require(_owner != address(0), "ShB owner 0");
+
+        eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber = _eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber;
+
+        l2BridgeAddress[ERA_CHAIN_ID] = ERA_ERC20_BRIDGE_ADDRESS;
+    }
+
+    /// @dev Initializes the l2Bridge address by governance for a specific chain
+    function initializeChainGovernance(uint256 _chainId, address _l2BridgeAddress) external onlyOwner {
+        l2BridgeAddress[_chainId] = _l2BridgeAddress;
+    }
+
+    /// @notice used by bridgehub to aquire mintValue. If l2Tx fails refunds are sent to refundrecipient on L2
+    /// we also use it to keep to track each chain's assets
+    function bridgehubDepositBaseToken(
+        uint256 _chainId,
+        address _prevMsgSender,
+        address _l1Token,
+        uint256 _amount
+    ) external payable onlyBridgehubOrEra(_chainId) {
+        require(_amount != 0, "4T"); // empty deposit amount
+
+        if (_l1Token == ETH_TOKEN_ADDRESS) {
+            require(msg.value == _amount, "L1SharedBridge: msg.value not equal to amount");
+        } else {
+            /// This breaks the _depositeFunds function, it returns 0, as our balance doesn't increase
+            /// This should not happen, this bridge only calls the Bridgehub if Eth is the baseToken or (in the future for wrapped base token deposits)
+            require(_prevMsgSender != address(this), "ShB calling itself");
+
```

Comment:
This checks create more confusion than contribute to security IMHO.

1) Why it is not needed for `_l1Token == ETH_TOKEN_ADDRESS`? The same case contract sends funds to itself.
2) if `_prevMsgSender == address(this)` then nothing changes as deposited value is equal to zero